### ISSUE TITLE
Indirect access to  `inline val` defined in `private object` causes  `getter (...) was not inlined` error

### ISSUE THREAD

**Issue by marcinzh** [scala/scala3#20513]
## Compiler version

3.3.3
3.4.1

## Minimized code

```Scala
// object Konst:
private object Konst:
  inline val K = 1
  // inline def K = 1


object Use:
  // OK
  def a: Int = Konst.K

  // OK
  inline def b: Int = Konst.K

  // Error
  def c: Int = b
```

https://scastie.scala-lang.org/3bXUpYzlQ1eMnQOrLAOLuA

## Output

```scala
[error] -- Error: /home/me/Bug.scala:17:15 
[error] 17 |  def c: Int = b
[error]    |               ^
[error]    |               getter K is declared as `inline`, but was not inlined
[error]    |
[error]    |               Try increasing `-Xmax-inlines` above 32
[error]    |----------------------------------------------------------------------------
[error]    |Inline stack trace
[error]    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[error]    |This location contains code that was inlined from Bug.scala:14
[error] 14 |  inline def b: Int = Konst.K
[error]    |                      ^^^^^^^
[error]     ----------------------------------------------------------------------------
[error] one error found
```

We can mask the error by either:
- removing `inline`s
- changing `object Konst` from `private` to public
- changing `val K` to `def K` 

Originally, `Konst` and `Use`  were defined in separate files.

## Expectation

Compile without error & apply inlining

**Comment by mbovel** [comment]
This issue was picked for the Scala Issue Spree of tomorrow, June 11th. @hamzaremmal, @SethTisue and @rochala will be working on it. If you have any further insights into the issue or guidance on how to fix it, please leave it here.

**Comment by SethTisue** [comment]
`-Vprint:all` shows

```
[[syntax trees at end of MegaPhase{pruneErasedDefs, uninitialized, inlinePatterns, vcInlineMethods, seqLiterals, intercepted, getters, specializeFunctions, specializeTuples, collectNullableFields, elimOuterSelect, resolveSuper, functionXXLForwarders, paramForwarding, genericTuples, letOverApply, arrayConstructors}]]
...
  @SourceFile("S.scala") final module class Use() extends Object() {
...
    private inline def b: Int = scala.compiletime.erasedValue[Int]
    def c: Int = Use.inline$Konst.K:Int
    def inline$Konst: Konst = Konst
  }
}
```

I don't have much experience with inlining, but `def inline$Konst: Konst = Konst` seems odd to me. it seems like the idea is to evade the `private`-ness of `Konst` with a forwarder, but it isn't accomplishing that because the type is `Konst` anyway?

but it might be a symptom rather than a cause. I guess let's enable tracing in inlining to get it talking about what it's doing? it might be useful to view the traces for the versions that work, as well as the trace for the version that goes wrong

**Comment by SethTisue** [comment]
Because `private` is present, the forwarder `inline$Konst` is introduced

**Comment by SethTisue** [comment]
Not 100% confident this is accurate, but the following picture seems to be emerging:

Both inlining and constant-folding are in play here. Constant-folding happens after inlining, in `FirstTransform`, which calls `TreeInfo#constToLiteral`. But `def c: Int = Use.inline$Konst.K:Int` cannot be constant-folded because `inline$Konst` isn't stable.

But we don't yet have an idea for a fix...

**Comment by SethTisue** [comment]
Recall that the semantics of `inline val` are very restrictive; inline vals must have constant literal types and be pure.

Maybe instead of waiting for constant-folding to go wrong, we can make the inliner actually inline `K`?

**Comment by SethTisue** [comment]
When you look at `Inliner` you see lots of code that seems intended to handle `val` and `def` the same (lots of `ValOrDefDef`), so we're puzzled why this plays out differently with `def`.

**Comment by SethTisue** [comment]
Jędrzej noticed that `Typer` has:

```
    if sym.isInlineMethod then rhsCtx.addMode(Mode.InlineableBody)
```

perhaps that's in play here? But we don't see where `InlineableBody` actually has any effect downstream?

**Comment by EugeneFlesselle** [comment]
I think the issue is not caused by the bridges added due to the `private` modifier, but about selecting an inline val on the result of a non-inline method.

Minimisation:
```scala
class Foo:
  inline val F = 1

def foo(): Foo = ???
val _ = foo().F
```

**Comment by hamzaremmal** [comment]
Indeed, the issue looks very similar to what we have been seeing during the spree. I will have to investigate it further.
